#**Experiment-3**

**AIM**: To construct a Machine learning model for 2-class prediction(related and unrelated) and to reduce the misclassification rate.


In [ ]:
# importing libraries
import pandas as pd
import numpy as np
import os
import re
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from nltk.corpus import stopwords

In [ ]:
train = pd.read_csv(r"C:\Users\Shreyansh\Capstone_files\final_train_data.csv") # train dataset

In [ ]:
test = pd.read_csv(r"C:\Users\Shreyansh\Capstone_files\final_test_data.csv")  # test dataset

In [ ]:
train.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'Headline', 'Body ID', 'articleBody',
       'jaccard_similarity', 'Headline_Nouns', 'articleBody_Nouns',
       'jaccard_similarity_nouns', 'Headline_POS', 'articleBody_POS', 'Stance',
       'glove_similarity', 'kl_divergence', 'ngram_overlap',
       'semantic_similarity', 'stance_base', 'LDA_Score'],
      dtype='object')

In [ ]:
test.columns

Index(['Unnamed: 0', 'Headline', 'Stance', 'articleBody', 'stance_cat',
       'jaccard_similarity', 'Headline_pos', 'articleBody_pos',
       'jaccard_similarity_nouns', 'Headline_Nouns',
       'articleBody_pos_lemmatized', 'Headline_pos_lemmatized', 'LDA_Score',
       'glove_similarity', 'kl_divergence', 'ngram_overlap',
       'semantic_similarity', 'stance_base'],
      dtype='object')

In [ ]:
test2 = test[['Headline', 'Stance',"articleBody"]]

In [ ]:
test_reslults = test2.groupby(['Headline','Stance']).count()

In [ ]:
test_reslults[:3]

articleBody
Headline                                           Stance                
ab de villiers wife danielle welcomed newborn d... agree                7
actor commits suicide actress tragic social sam... unrelated            1
bitcoin tops wins mainstream acceptance            agree                1

In [ ]:
test_reslults2 = pd.DataFrame(test_reslults[:3])

In [ ]:
test_reslults2 = test_reslults2.reset_index()

In [ ]:
unrelated = test_reslults2[test_reslults2.Stance == "agree"]
unrelated

Headline Stance  articleBody
0  ab de villiers wife danielle welcomed newborn ...  agree            7
2            bitcoin tops wins mainstream acceptance  agree            1

In [ ]:
unrelated["articleBody"].iloc[0]

7

In [ ]:
train["stance_base"].unique()

array(['unrelated', 'related'], dtype=object)

In [ ]:
test["stance_base"].unique()

array(['related', 'unrelated'], dtype=object)

In [ ]:
import itertools   # consider 7-features as predictors
a = []
stuff = ['jaccard_similarity', 'jaccard_similarity_nouns', 'glove_similarity', 'kl_divergence', 'ngram_overlap',
         "semantic_similarity", "LDA_Score"]
for L in range(0, len(stuff)+1):
    for subset in itertools.combinations(stuff, L):
        a.append(list(subset))
        
print(len(a))
#print(a)

x_ = a
y = len(x_)
xyz = x_[-(y-1):]
print(xyz)

128
[['jaccard_similarity'], ['jaccard_similarity_nouns'], ['glove_similarity'], ['kl_divergence'], ['ngram_overlap'], ['semantic_similarity'], ['LDA_Score'], ['jaccard_similarity', 'jaccard_similarity_nouns'], ['jaccard_similarity', 'glove_similarity'], ['jaccard_similarity', 'kl_divergence'], ['jaccard_similarity', 'ngram_overlap'], ['jaccard_similarity', 'semantic_similarity'], ['jaccard_similarity', 'LDA_Score'], ['jaccard_similarity_nouns', 'glove_similarity'], ['jaccard_similarity_nouns', 'kl_divergence'], ['jaccard_similarity_nouns', 'ngram_overlap'], ['jaccard_similarity_nouns', 'semantic_similarity'], ['jaccard_similarity_nouns', 'LDA_Score'], ['glove_similarity', 'kl_divergence'], ['glove_similarity', 'ngram_overlap'], ['glove_similarity', 'semantic_similarity'], ['glove_similarity', 'LDA_Score'], ['kl_divergence', 'ngram_overlap'], ['kl_divergence', 'semantic_similarity'], ['kl_divergence', 'LDA_Score'], ['ngram_overlap', 'semantic_similarity'], ['ngram_overlap', 'LDA_Score'

In [ ]:
train['stance_base'] 

0        unrelated
1          related
2        unrelated
3        unrelated
4          related
           ...    
75380      related
75381      related
75382      related
75383      related
75384      related
Name: stance_base, Length: 75385, dtype: object

In [ ]:
train["Stance"].unique()

array(['unrelated', 'agree', 'disagree', 'discuss'], dtype=object)

In [ ]:
test["Stance"].unique()

array(['agree', 'discuss', 'disagree', 'unrelated'], dtype=object)

In [ ]:
#train =  train[(train.Stance == 'disagree') | (train.Stance == 'discuss')]

In [ ]:
train["Stance"].unique()

array(['unrelated', 'agree', 'disagree', 'discuss'], dtype=object)

In [ ]:
#test =  test[(test.Stance == 'disagree') | (test.Stance == 'discuss')]

test["Stance"].unique()

array(['agree', 'discuss', 'disagree', 'unrelated'], dtype=object)

In [ ]:
def combinations_test_data(combi_list):    # creat a function to generate 49 combination feature from 7-features
    score = []
    for i in combi_list:
        from sklearn.metrics import accuracy_score
        from sklearn.model_selection import train_test_split
        from sklearn.ensemble import RandomForestClassifier
        from nltk.corpus import stopwords


        #xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.1, random_state = 101)
        xtrain3 = train[i]
        ytrain3 = train['stance_base'] 
        xtest3 = test[i]
        ytest3 = test["stance_base"]
        rg = RandomForestClassifier(n_estimators=100,n_jobs=-1)


        r = len(i)
        #print('X Training shape',xtrain.shape)
        #print('Y Training shape',ytrain.shape)
        xtrain3 = xtrain3.values.reshape(-1,r)
        xtest3 = xtest3.values.reshape(-1,r)

        rg.fit(xtrain3,ytrain3)
        ypred2 = rg.predict(xtest3)
        print('Accuracy score on three class ',accuracy_score(ypred2,ytest3))
        print(i)

        #ypred2[:20]

        from sklearn.metrics import classification_report, confusion_matrix

        print(classification_report(ypred2,ytest3))
        print(confusion_matrix(ypred2,ytest3))
        score.append(accuracy_score(ypred2,ytest3))
    print("max accuracy" ,max(score))
    print(score)


In [ ]:
combinations_test_data(combi_list = xyz)

Accuracy score on three class  0.7735849056603774
['jaccard_similarity']
              precision    recall  f1-score   support

     related       0.87      0.79      0.83       145
   unrelated       0.62      0.75      0.68        67

    accuracy                           0.77       212
   macro avg       0.74      0.77      0.75       212
weighted avg       0.79      0.77      0.78       212

[[114  31]
 [ 17  50]]
Accuracy score on three class  0.8018867924528302
['jaccard_similarity_nouns']
              precision    recall  f1-score   support

     related       0.84      0.84      0.84       131
   unrelated       0.74      0.74      0.74        81

    accuracy                           0.80       212
   macro avg       0.79      0.79      0.79       212
weighted avg       0.80      0.80      0.80       212

[[110  21]
 [ 21  60]]
Accuracy score on three class  0.5707547169811321
['glove_similarity']
              precision    recall  f1-score   support

     related       0.4

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Accuracy score on three class  0.3867924528301887
['semantic_similarity']
              precision    recall  f1-score   support

     related       0.13      0.52      0.21        33
   unrelated       0.80      0.36      0.50       179

    accuracy                           0.39       212
   macro avg       0.47      0.44      0.35       212
weighted avg       0.70      0.39      0.45       212

[[ 17  16]
 [114  65]]
Accuracy score on three class  0.37264150943396224
['LDA_Score']
              precision    recall  f1-score   support

     related       0.01      0.25      0.01         4
   unrelated       0.96      0.38      0.54       208

    accuracy                           0.37       212
   macro avg       0.49      0.31      0.28       212
weighted avg       0.94      0.37      0.53       212

[[  1   3]
 [130  78]]
Accuracy score on three class  0.7311320754716981
['jaccard_similarity', 'jaccard_similarity_nouns']
              precision    recall  f1-score   support

     

Accuracy score on three class  0.6698113207547169
['kl_divergence', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.79      0.71      0.75       145
   unrelated       0.48      0.58      0.53        67

    accuracy                           0.67       212
   macro avg       0.63      0.65      0.64       212
weighted avg       0.69      0.67      0.68       212

[[103  42]
 [ 28  39]]
Accuracy score on three class  0.38207547169811323
['ngram_overlap', 'semantic_similarity']
              precision    recall  f1-score   support

     related       0.21      0.50      0.29        54
   unrelated       0.67      0.34      0.45       158

    accuracy                           0.38       212
   macro avg       0.44      0.42      0.37       212
weighted avg       0.55      0.38      0.41       212

[[ 27  27]
 [104  54]]
Accuracy score on three class  0.37264150943396224
['ngram_overlap', 'LDA_Score']
              precision    recall  f1-score  

Accuracy score on three class  0.7735849056603774
['jaccard_similarity_nouns', 'glove_similarity', 'kl_divergence']
              precision    recall  f1-score   support

     related       0.91      0.77      0.83       155
   unrelated       0.56      0.79      0.65        57

    accuracy                           0.77       212
   macro avg       0.73      0.78      0.74       212
weighted avg       0.81      0.77      0.78       212

[[119  36]
 [ 12  45]]
Accuracy score on three class  0.8066037735849056
['jaccard_similarity_nouns', 'glove_similarity', 'ngram_overlap']
              precision    recall  f1-score   support

     related       0.88      0.82      0.85       140
   unrelated       0.69      0.78      0.73        72

    accuracy                           0.81       212
   macro avg       0.78      0.80      0.79       212
weighted avg       0.81      0.81      0.81       212

[[115  25]
 [ 16  56]]
Accuracy score on three class  0.6084905660377359
['jaccard_similari

Accuracy score on three class  0.4339622641509434
['kl_divergence', 'semantic_similarity', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.30      0.58      0.39        67
   unrelated       0.65      0.37      0.47       145

    accuracy                           0.43       212
   macro avg       0.48      0.47      0.43       212
weighted avg       0.54      0.43      0.45       212

[[39 28]
 [92 53]]
Accuracy score on three class  0.38207547169811323
['ngram_overlap', 'semantic_similarity', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.20      0.50      0.28        52
   unrelated       0.68      0.34      0.46       160

    accuracy                           0.38       212
   macro avg       0.44      0.42      0.37       212
weighted avg       0.56      0.38      0.41       212

[[ 26  26]
 [105  55]]
Accuracy score on three class  0.75
['jaccard_similarity', 'jaccard_similarity_nouns', 'glove_s

Accuracy score on three class  0.6792452830188679
['jaccard_similarity', 'kl_divergence', 'ngram_overlap', 'semantic_similarity']
              precision    recall  f1-score   support

     related       0.84      0.70      0.76       157
   unrelated       0.42      0.62      0.50        55

    accuracy                           0.68       212
   macro avg       0.63      0.66      0.63       212
weighted avg       0.73      0.68      0.70       212

[[110  47]
 [ 21  34]]
Accuracy score on three class  0.7264150943396226
['jaccard_similarity', 'kl_divergence', 'ngram_overlap', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.91      0.72      0.80       165
   unrelated       0.43      0.74      0.55        47

    accuracy                           0.73       212
   macro avg       0.67      0.73      0.68       212
weighted avg       0.80      0.73      0.75       212

[[119  46]
 [ 12  35]]
Accuracy score on three class  0.7216981132075472


Accuracy score on three class  0.419811320754717
['kl_divergence', 'ngram_overlap', 'semantic_similarity', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.34      0.55      0.42        80
   unrelated       0.56      0.34      0.42       132

    accuracy                           0.42       212
   macro avg       0.45      0.45      0.42       212
weighted avg       0.47      0.42      0.42       212

[[44 36]
 [87 45]]
Accuracy score on three class  0.7311320754716981
['jaccard_similarity', 'jaccard_similarity_nouns', 'glove_similarity', 'kl_divergence', 'ngram_overlap']
              precision    recall  f1-score   support

     related       0.92      0.72      0.81       168
   unrelated       0.42      0.77      0.54        44

    accuracy                           0.73       212
   macro avg       0.67      0.75      0.68       212
weighted avg       0.82      0.73      0.75       212

[[121  47]
 [ 10  34]]
Accuracy score on three class

Accuracy score on three class  0.7688679245283019
['jaccard_similarity_nouns', 'glove_similarity', 'kl_divergence', 'ngram_overlap', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.89      0.77      0.83       150
   unrelated       0.58      0.76      0.66        62

    accuracy                           0.77       212
   macro avg       0.73      0.77      0.74       212
weighted avg       0.80      0.77      0.78       212

[[116  34]
 [ 15  47]]
Accuracy score on three class  0.7216981132075472
['jaccard_similarity_nouns', 'glove_similarity', 'kl_divergence', 'semantic_similarity', 'LDA_Score']
              precision    recall  f1-score   support

     related       0.79      0.76      0.78       136
   unrelated       0.60      0.64      0.62        76

    accuracy                           0.72       212
   macro avg       0.70      0.70      0.70       212
weighted avg       0.73      0.72      0.72       212

[[104  32]
 [ 27  49]]
Ac

In [ ]:
['jaccard_similarity_nouns', 'glove_similarity']

['jaccard_similarity_nouns', 'glove_similarity']

In [ ]:
['jaccard_similarity_nouns', 'glove_similarity', 'ngram_overlap']

In [ ]:
# import libraries
    
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from nltk.corpus import stopwords


In [ ]:
 #xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.1, random_state = 101)
xtrain3 = train[['jaccard_similarity_nouns', 'glove_similarity', 'ngram_overlap']]
ytrain3 = train['stance_base'] 
xtest3 = test[['jaccard_similarity_nouns', 'glove_similarity', 'ngram_overlap']]
ytest3 = test["stance_base"]
rg = RandomForestClassifier(n_estimators=100,n_jobs=-1)

In [ ]:
xtrain3 = xtrain3.values.reshape(-1,3)
xtest3 = xtest3.values.reshape(-1,3)

rg.fit(xtrain3,ytrain3)
ypred2 = rg.predict(xtest3)

In [ ]:
train.shape

(75385, 18)

In [ ]:
test.shape

(212, 18)

In [ ]:

print('Accuracy score on two class related and unrelated class ',accuracy_score(ypred2,ytest3))
from sklearn.metrics import classification_report, confusion_matrix
print(classification_report(ypred2,ytest3))
print(confusion_matrix(ypred2,ytest3))3



Accuracy score on two class related and unrelated class  0.8113207547169812
              precision    recall  f1-score   support

     related       0.89      0.82      0.85       141
   unrelated       0.69      0.79      0.74        71

    accuracy                           0.81       212
   macro avg       0.79      0.81      0.79       212
weighted avg       0.82      0.81      0.81       212

[[116  25]
 [ 15  56]]


**Results :  The best combination of features turned out to be Jaccard nouns similarity, GloVe similarity and N-grams. These 3 features gave 81% accuracy using Random forest classifier.**